### PR TITLE
 [#10179] Create invite link for group chat

### LINKED ISSUE #10179: Create invite link for group chat
# Feature Issue
Create an invite link for group chats that can be used as a referral link

### User Story

- As a user I want to invite friends in my current group chat to join me on Status in a private group chat, because I want to continue our existing shared conversation there
- As Status we want to onboard existing groups as they are more likely to stay compared to individuals being connected to one other user / not interested in public group chats
- As a user who sent out a link I want to be part of the referral scheme using that link so that I can earn SNTs

Deep linking throughout install only works on Android. on iOS recipients of an invite link will always need to reopen the link after installing the app.

##### User requirements
- Admins can create a group chat link
- Anyone can share the link once it's created by an admin
- The admin who created the link gets notified to Approve/Reject
_*Only admins as the link needs to include their pk in order for them to Approve joining_

##### Flow
- User B wants to invite someone, they tell user A
- User A creates the invite link url (note that it will be long, includes a group chat ID, chat name, admin pk)
- User B invites someone
- User A sees a notification to approve

### Notes
Effort estimate 3 weeks upon availability of designs

### PR DESCRIPTION
**Pull request by flexsurfer**
fixes #10179

* [x] generate link UI
* [x] open link UI
* [x] send request message
* [x] receive and accept / decline request

### CODE REVIEW

**Comment by hesterbruikman** [comment]
@flexsurfer Noticing the two open tasks:
- [ ] send request message
- [ ] receive and accept / decline request

Are you blocked by anything? Still think this is relevant. Could probably use the flow for moderated public chats (organization channels) lateron as well. Designs got a little lost in the thread, but are [here](https://www.figma.com/file/aS1ct66VQ6V0cio7vSqS8UoG/Chat?node-id=8927%3A39804)

Let me know if you have any doubts about the work

**Comment by flexsurfer** [comment]
@hesterbruikman no i'm not blocking anymore and i keep working on it, i don't have experience with go and messages, so it will take some time for me

**Comment by hesterbruikman** [comment]
Fantastic work @flexsurfer! Had a true moment of delight seeing a familiar group right after onboarding!

#### Use case 1: User A (admin) creates invite link on Android. User B opens link on iOS 
_(Status already downloaded, onboarding not completed, can't test join.status flow yet)_
Result: @flexsurfer rocks. Group chat is listed in chat overview, `membership request` can be send. 

#### Use case 2: User B sends membership request
Result: Group admin sees group membership request
- Issue 1: Input field hidden under Keyboard on iOS (OK on Android)
- Issue 2: Character limit validation not informative. Can exceed and send request. Not clear if text will be cut off. Expected: @errorists ?
- Issue 3: There is currently no local, push or badge notification that informs User A that there is a Group membership request
- Issue 4: Group invite view has label `Pending invitations`. Expected: `Pending membership requests` @errorists wdyt, as far as I can tell sent invites currently do not show up, pending membership requests do. 

#### Use case 3: User A (admin) accepts membership request
Result: User B gets group chat invite and can join as per current invite
- ✅ No issues

#### Use case 4: User A (admin) declines membership request
Result: User B gets `membership request was declined`
- ✅ No issues

#### Use case 5: User B retries membership request
Result: User A sees `Group membership request again`
- ✅ No issues

#### Use case 6: User B removes group
Result: User A sees `Group membership request again`
- Issue 5: On tap `Remove group` nothing happens, button responds to tap, but no action (Android and iOS). Expected: Group view closes, user returns to chat overview list without Group chat listed

#### Use case 7: User B opens invite link when Status is installed and account is open
Result: App is opened in Group chat membership request view
- ✅ No issues

#### Use case 8: User B opens invite link when Status is installed, account is locked
Result: App is opened, after unlock, Group chat membership request view opens
- ✅ No issues

#### Use case 9: User B opens invite link when Status is installed, app not running
Result: App is opened, after unlock, Group chat membership request view opens
- ✅ No issues

#### Use case 10: User B opens link with install that already has an existing multiaccount (open or not)
- Issue 6: On Android 7, group does not appear in chat list. No issues on Android 10 and iOS. Expected: Group chat appears in chat list when:
     - new user onboards, 
     - user has an existing multiaccount (open or closed)
     - had a multiaccount, removed and created a new multiaccount

**Comment by flexsurfer** [comment]
thanks @hesterbruikman , Issue 2: and Issue 5 have been fixed

**Comment by flexsurfer** [comment]
not sure how to reproduce 
Issue 1: Input field hidden under Keyboard on iOS (OK on Android)
and
Issue 6: On Android 7, group does not appear in chat list.

**Comment by hesterbruikman** [comment]
@flexsurfer regarding issue 1, here's a [recording](https://drive.google.com/file/d/1h1FUSYHEz6Hp8-vlVeNuGPYW7mq6anka/view?usp=sharing) of iPhone 8. @errorists can you see if it's reproducible on a newer model

**Comment by errorists** [comment]
oh I see what you mean now, that's not good. @flexsurfer shouldn't KeyboardAvoidingView help here to scroll-focus the input in the center of the view?

**Comment by flexsurfer** [comment]
thanks, i can repro now , will fix

**Comment by flexsurfer** [comment]
@errorists what about issue 4 ?

**Comment by errorists** [comment]
@flexsurfer I think it's a good request, so let's go with `Pending membership requests`

**Comment by flexsurfer** [comment]
fixed issue 1 and issue 4

**Comment by hesterbruikman** [comment]
For reference, these issues are open, correct @flexsurfer ?
- Issue 3: There is currently no local, push or badge notification that informs User A that there is a Group membership request
- Issue 6: On Android 7, group does not appear in chat list. No issues on Android 10 and iOS. Expected: Group chat appears in chat list when:
   - new user onboards,
   - user has an existing multiaccount (open or closed)
   - had a multiaccount, removed and created a new multiaccount

Issue 3 should probably be fixed outside of this PR, as part of notifications (currently we have no notifications for group chats on Android)
Issue 6 I think needs more investigation by QA to check if this occurs on other versions of Android

**Comment by flexsurfer** [comment]
yes, correct , thanks, agree PN outside of this PR, and Issue 6 looks like a general issue for universal links not specific only for this PR

**Comment by churik** [comment]
@flexsurfer 
testing is still in progress, issues/questions:
### QUESTION 7: no indicators of pending membership request on home view
I would like to have some badge in home view informed that I have some unresponded group membership request
@errorists what do you think?


### QUESTION 8: Universal QR scanner is not working with group invites
IMO should be done in this PR as well
**Steps:**
- User A: create group chat and make invite link
- User B: scan QR code with invite link via Status > + > universal scanner
( `status-im://g/...`, ` https://join.status.im/g/..`)

### ISSUE 9: Message is cut in "Request Pending" state
**Steps:**
- User A: create group chat and make invite link
- User B: open invite link and set message with 100 chars
- User B: Request access
photo_2020-08-18 17 27 46 jpeg 2020-08-18 17-28-16


### ISSUE 10: Accepted membersip request reappears after relogin
**Steps:**
- User A: create group chat and make invite link
- User B: open invite link, request access
- User A: accept request
- User B: join group chat
- User A: reopen app and open group chat
**Expected result:** no pending 'Group membership request'
**Actual result:** pending 'Group membership request' from `User A`


### ISSUE 11:  'Group membership request' re-created if navigating to invite several times
**Steps:**
- User A: create group chat and make invite link
- User B: open invite link, request access
- User A: accept request
- User B: join group chat
- User B: open invite link one more time
**Expected result:** will be redirected to group chat, no new request to join
**Actual result:** 'Request pending' state in group chat

**Comment by flexsurfer** [comment]
thanks, fixed 8,9,11
7 needs response from Maciej
10 needs help from @cammellos , for some reason invitation message is received again

File: src/status_im/chat/models.cljs [hunk 1]
```diff
@@ -110,10 +110,12 @@
...
-    (merge
-     (or (get (:chats db) (:chat-id val))
-         (create-new-chat (:chat-id val) cofx))
-     val)))
+    (assoc
+     (merge
+      (or (get (:chats db) (:chat-id val))
+          (create-new-chat (:chat-id val) cofx))
+      val)
+     :invitation-admin (:invitation-admin val))))
```
Comment (Ferossgp) [inline src/status_im/chat/models.cljs:118]: Looks like it is redundant as the whole `val` is merged above?

File: src/status_im/data_store/invitations.cljs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+(ns status-im.data-store.invitations
+  (:require clojure.set))
+
+(defn <-rpc [message]
```
Comment (cammellos) [inline src/status_im/data_store/invitations.cljs:4]: Destructuring is much faster, and probably just as readable
```
(def a {:chatId "a"  :introductionMessage "message" :messageType "message-type"})

(defn <-rpc-rename [message]
  (-> message
      (clojure.set/rename-keys {:chatId              :chat-id
                                :introductionMessage :introduction-message
                                :messageType         :message-type})))

(defn <-rpc-destruct [{:keys [chatId introductionMessage messageType]}]
  {:chat-id chatId
   :introduction-message introductionMessage
   :message-type messageType})


(defn time-rename []
  (time
   (doseq [_ (range 10000)]
     (<-rpc-rename a))))

(defn time-destruct []
  (time
   (doseq [_ (range 10000)]
     (<-rpc-destruct a))))

cljs.user=> (time-rename)
"Elapsed time: 68.831563 msecs"
nil
cljs.user=> (time-destruct)
"Elapsed time: 14.788217 msecs"

```
destructring is around 4/5 times faster, probably does not matter much, but I would just use destructuring here.

**Comment by errorists** [comment]
re: 7
we [had those in the group chat designs](https://www.figma.com/file/aS1ct66VQ6V0cio7vSqS8UoG/Chat?node-id=480%3A632). basically, for any chat on the list that asks your attention so: a new private chat, new group invite, we should change their background colour to light blue, for groups until you choose accept or decline or was accepted or declined into a group for links.

File: src/status_im/router/core.cljs [hunk 3]
```diff
@@ -87,6 +92,22 @@
     {:type  :public-chat
      :error :invalid-topic}))
 
+(defn match-group-chat [uri _]
+  ;;TODO there is a bug in bidi, so we have to parse uri manually here
+  (let [[_ params] (when (not (string/blank? uri)) (string/split uri #"/g/"))
+        [admin-pk encoded-chat-name chat-id]  (when (not (string/blank? params)) (string/split params #"&"))
```
Comment (Ferossgp) [inline src/status_im/router/core.cljs]: This looks to me like not a standard Query string. I think using query params for that could be easier for multiple clients. Like `?admin-pk=0x0&encoded-chat-name=abcd&chat-id=abcd` https://en.wikipedia.org/wiki/Query_string#Structure

Also bidi does not have problems with standard query params.

Also, Query Strings may be easier to handle on `join.status.im`
cc. @jakubgs

**Review by cammellos — APPROVED** [review]
looks good, there's some outstanding comments on the status-go PR, I would also add so tests to `match-group-chat` as there's quite a lot of logic in that fn

**Review by rasom — DISMISSED** [review]
@flexsurfer can you please respond to existing comments from @Ferossgp and @cammellos?

File: src/status_im/router/core.cljs (hunk 3 above)
Comment (jakubgs) [reply src/status_im/router/core.cljs]: What are you asking me though?

Comment (Ferossgp) [reply src/status_im/router/core.cljs]: The current link format is using `&` to separate list of params.
```
https://join.status.im/g/0x0454a896b7b3307e402fa1ae78cb5c5dda27346f63c1581bbdc91eb3bed76253f367801a4fa44ee5ebd48c5e5210b69faea5671a26568e6fb2138fe6f1c4fb4bec&hi&63f5c71a-86b8-492a-a215-3dcca72fddbb-0x0454a896b7b3307e402fa1ae78cb5c5dda27346f63c1581bbdc91eb3bed76253f367801a4fa44ee5ebd48c5e5210b69faea5671a26568e6fb2138fe6f1c4fb4bec
```
While I was thinkig on this format.
```
https://join.status.im/g?admin-k=0x0454a896b7b3307e402fa1ae78cb5c5dda27346f63c1581bbdc91eb3bed76253f367801a4fa44ee5ebd48c5e5210b69faea5671a26568e6fb2138fe6f1c4fb4bec&chat-name=hi&chat-id=63f5c71a-86b8-492a-a215-3dcca72fddbb-0x0454a896b7b3307e402fa1ae78cb5c5dda27346f63c1581bbdc91eb3bed76253f367801a4fa44ee5ebd48c5e5210b69faea5671a26568e6fb2138fe6f1c4fb4bec
```

Asking you cause I know that you do the changes on unviersal link websites usually

Comment (jakubgs) [reply src/status_im/router/core.cljs]: Damn that's a long-ass URL.

I'd say using proper URL arguments is cleaner, though wondering if some part of it should be part of the path?

Though I'm kinda confused. I thought referral program arguments will be passed via URL arguments for normal channel/chat links, like:
```
https://join.status.im/some-channel-name?some-referral-garbage=1231231231231231231231231231231232131
```
or
```
https://join.status.im/u/some-username?some-referral-garbage=1231231231231231231231231231231232131
```
But now you're showing me something completely new, what is `/g` supposed to be?

Comment (Ferossgp) [reply src/status_im/router/core.cljs]: This PR is about Group Chats invite (By this link user can join a group chat), not related to referrals

Comment (Ferossgp) [reply src/status_im/router/core.cljs]: > though wondering if some part of it should be part of the path

Yeah, I think maybe `chat-id` could be a path/route param

Comment (flexsurfer) [reply src/status_im/router/core.cljs]: yeah i thought about that as well, will change it to query params

Comment (jakubgs) [reply src/status_im/router/core.cljs]: I think that's better, it makes the most important part of the URL part of the path, and the rest be parameters.

File: src/status_im/chat/models.cljs (hunk 1 above)
Comment (flexsurfer) [reply src/status_im/chat/models.cljs:118]: not its not, this is for `nil`

File: src/status_im/data_store/invitations.cljs (hunk 2 above)
Comment (flexsurfer) [reply src/status_im/data_store/invitations.cljs:4]: thanks , wondering why

Comment (flexsurfer) [reply src/status_im/data_store/invitations.cljs:4]: thanks

Comment (flexsurfer) [reply src/status_im/data_store/invitations.cljs:4]: but you create a new map, but i need to remove keys and assign new keys

Comment (flexsurfer) [reply src/status_im/data_store/invitations.cljs:4]: also this approach is general for the app, not just introduced in this PR

Comment (cammellos) [reply src/status_im/data_store/invitations.cljs:4]: up to you, it's true that we use rename-keys more often, maybe it's time to move away from it :), but either are fine with me

**Comment by flexsurfer** [comment]
@churik fixed 10 and 11

**Comment by churik** [comment]
@flexsurfer sorry for delay!
### QUESTION 7: no indicators of pending membership request on home view
Not resolved according to https://github.com/status-im/status-react/pull/10937#issuecomment-678956384

### ISSUE 12: user is redirected to home view instead membership request after onboarding (when new multiaccount is created)
After tapping on valid `https://join.status.im/g/..` invite link and navigate to app using 'Generate a new key' flow, user will end up on home view instead of redireting to group chat view with membership request.
All fine if using flow `Access existing keys`



### ISSUE 13: List in 'Group invite' is not scrollable
If Group invites > 4 (tested on IPhone 7), last 3 items are unaccessible
Monosnap 2020-09-01 12-48-53

### ISSUE 14: Limit of 10 members is not working with group invites
As the number of membership requests is not limited, you can accept any amount of requests and exceed the limit of 10 members.
IMAGE 2020-09-01 12:54:35


### ISSUE 15: Can click 'Join',  'Accept' several times 
Consequense:
IMAGE 2020-09-01 13:01:17

**Comment by flexsurfer** [comment]
ISSUE 14: Limit of 10 members is not working with group invites
@errorists could you please elaborate how do you see it ? do we want to disable list and show note that its not possible to add  more ?

**Comment by flexsurfer** [comment]
thanks @churik 13-15 have been fixed, QUESTION 7 and ISSUE 12 won't be addressed in this PR

**Pull request merged**